kernel/aster-nix/src/vdso.rs

```diff
@@ -197,51 +231,90 @@ impl Vdso {
         }
     }
 
-    /// Return the vdso vmo.
+    /// Return the VDSO vmo.
     fn vmo(&self) -> Arc<Vmo> {
         self.vmo.clone()
     }
 
-    fn update_instant(&self, instant: Instant, instant_cycles: u64) {
-        self.data.lock().update_instant(instant, instant_cycles);
+    fn update_high_res_instant(&self, instant: Instant, instant_cycles: u64) {
+        self.data
+            .lock()
+            .update_high_res_instant(instant, instant_cycles);
 
         // Update begins.
+        let seq_lock = SEQ_LOCK.lock();
         self.vmo.write_val(0x80, &1).unwrap();
         self.vmo.write_val(0x88, &instant_cycles).unwrap();
-        for clock_id in ALL_SUPPORTED_CLOCK_IDS {
+        for clock_id in HIGH_RES_CLOCK_IDS {
             self.update_vmo_instant(clock_id);
         }
         // Update finishes.
         self.vmo.write_val(0x80, &0).unwrap();
     }
 
-    /// Update the requisite fields of the vdso data in the vmo.
+    fn update_coarse_res_instant(&self, instant: Instant) {
+        self.data.lock().update_coarse_res_instant(instant);
+
+        // Update begins.
+        let seq_lock = SEQ_LOCK.lock();
+        self.vmo.write_val(0x80, &1).unwrap();
+        for clock_id in COARSE_RES_CLOCK_IDS {
+            self.update_vmo_instant(clock_id);
+        }
+        // Update finishes.
+        self.vmo.write_val(0x80, &0).unwrap();
+    }
+
+    /// Update the requisite fields of the VDSO data in the vmo.
     fn update_vmo_instant(&self, clockid: ClockID) {
         let clock_index = clockid as usize;
         let secs_offset = 0xA0 + clock_index * 0x10;
-        let nanos_lshift_offset = 0xA8 + clock_index * 0x10;
+        let nanos_info_offset = 0xA8 + clock_index * 0x10;
         let data = self.data.lock();
         self.vmo
             .write_val(secs_offset, &data.basetime[clock_index].secs)
             .unwrap();
         self.vmo
-            .write_val(
-                nanos_lshift_offset,
-                &data.basetime[clock_index].nanos_lshift,
-            )
+            .write_val(nanos_info_offset, &data.basetime[clock_index].nanos_info)
             .unwrap();
     }
 }
 
-/// Update the `VdsoInstant` in Vdso.
-fn update_vdso_instant(instant: Instant, instant_cycles: u64) {
-    VDSO.get().unwrap().update_instant(instant, instant_cycles);
+/// Update the `VdsoInstant` for clock IDs with high resolution in Vdso.
+fn update_vdso_high_res_instant(instant: Instant, instant_cycles: u64) {
+    let work_func = Box::new(move || {
+        VDSO.get()
+            .unwrap()
+            .update_high_res_instant(instant, instant_cycles);
+    });
+
+    // Updating involves writing operations to VMO, which will lock a `Mutex` and
+    // should be put in the work queue.
+    let work_item = Arc::new(WorkItem::new(work_func));
+    submit_work_item(work_item, crate::thread::work_queue::WorkPriority::High);
```

Comment:
Could you try to avoid using a work queue to update the VDSO data? This approach is too heavy weight for such a frequent and simple operation. In addition, work items may be processed by work queues in arbitrary latencies. This is especially true considering all current task schedulers are naive and non-preemptive. In fact, even if our schedulers were good, we should NOT let updating time depend on task scheduling because scheduling itself must depend on accurate time accounting.

One more efficient approach to updating the VDSO data is to use `VmFrame` instead of `Vmo`. This can be done if we add some methods to `Vmo` that allows obtaining the backing `VmFrame`s of a `Vmo`.